types/src/shared/utils/structured_data.ts

```diff
@@ -27,3 +31,51 @@ export function getSanitizedHeaders(rawHeaders: string[]) {
     return acc;
   }, []);
 }
+
+export async function guessDelimiter(csv: string): Promise<string | undefined> {
+  // Detect the delimiter: try to parse the first 2 lines with different delimiters,
+  // keep the one that works for both lines and has the most columns.
+  let delimiter: string | undefined = undefined;
+  let delimiterColsCount = 0;
+  for (const d of [",", ";", "\t"]) {
+    const records: unknown[][] = [];
+    try {
+      const parser = parse(csv, { delimiter: d });
+      for await (const record of parser) {
+        records.push(record);
+        if (records.length == 2) {
+          break;
+        }
+      }
+    } catch (e) {
+      // Ignore error.
+      continue;
+    }
+
+    const [firstRecord, secondRecord] = records;
+    // Check for more than one line to ensure sufficient data for accurate delimiter detection.
+    if (!secondRecord) {
+      continue;
+    }
+
+    if (!!firstRecord.length && firstRecord.length === secondRecord.length) {
+      if (firstRecord.length > delimiterColsCount) {
+        delimiterColsCount = firstRecord.length;
+        delimiter = d;
+      }
+    }
+  }
+
+  return delimiter;
+}
+
+export async function parseAndStringifyCsv(tableCsv: string) {
+  const delimiter = await guessDelimiter(tableCsv);
+  const records = parse_sync(tableCsv, {
```

Comment:
why are we using the synchronous one here ? @flvndvd  since you added it. If the file is big, this will block the event loop.

Reply:
IIRC, this used to be called on the client side so it was not a big deal, but agree that if we use it on the server side as well, we should use the async version.

Reply:
ah yes I think you're right 👍  Client side is fine but not server sice

Reply:
since this is now in type I would move to async because there's no guarantee this will remain client side

Reply:
Everything should be async now 👍🏼